tensorboard/plugins/beholder/video_writing.py

```diff
@@ -12,4 +12,189 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-# TODO(@jart): Rewrite me.
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import abc
+import os
+import subprocess
+import time
+
+import numpy as np
+import tensorflow as tf
+
+from tensorboard.plugins.beholder import im_util
+
+
+class VideoWriter(object):
+  """Video file writer that can use different output types.
+
+  Each VideoWriter instance writes video files to a specified directory, using
+  the first available VideoOutput from the provided list.
+  """
+
+  def __init__(self, directory, outputs):
+    self.directory = directory
+    # Filter to the available outputs
+    self.outputs = [out for out in outputs if out.available()]
+    if not self.outputs:
+      raise IOError('No available video outputs')
+    self.output_index = 0
+    self.output = None
+    self.frame_shape = None
+
+  def current_output(self):
+    return self.outputs[self.output_index]
+
+  def write_frame(self, np_array):
+    # Reset whenever we encounter a new frame shape.
+    if self.frame_shape != np_array.shape:
+      if self.output:
+        self.output.close()
+      self.output = None
+      self.frame_shape = np_array.shape
+      tf.logging.info('Starting video with frame shape: %s', self.frame_shape)
+    # Write the frame, advancing across output types as necessary.
+    original_output_index = self.output_index
+    for self.output_index in range(original_output_index, len(self.outputs)):
+      try:
+        if not self.output:
+          new_output = self.outputs[self.output_index]
+          if self.output_index > original_output_index:
+            tf.logging.warn(
+                'Falling back to video output %s', new_output.name())
+          self.output = new_output(self.directory, self.frame_shape)
+        self.output.emit_frame(np_array)
+        return
+      except (IOError, OSError) as e:
+        tf.logging.warn(
+            'Video output type %s not available: %s',
+            self.current_output().name(), str(e))
+        if self.output:
+          self.output.close()
+        self.output = None
+    raise IOError('Exhausted available video outputs')
+
+  def finish(self):
+    if self.output:
+      self.output.close()
+    self.output = None
+    self.frame_shape = None
+    # Reconsider failed outputs when video is manually restarted.
+    self.output_index = 0
+
+
+class VideoOutput(object):
+  """Base class for video outputs supported by VideoWriter."""
+
+  __metaclass__ = abc.ABCMeta
+
+  # Would add @abc.abstractmethod in python 3.3+
+  @classmethod
+  def available(cls):
+    raise NotImplementedError()
+
+  @classmethod
+  def name(cls):
+    return cls.__name__
+
+  @abc.abstractmethod
+  def emit_frame(self, np_array):
+    raise NotImplementedError()
+
+  @abc.abstractmethod
+  def close(self):
```

Comment:
nit: Should we comment on how close doesn't immediately terminate the writing, but rather waits for writing to complete (and obtains and stdout and stderr logs)? I think that's what the call to `communicate` within `FFmpegVideoOutput`'s `close` method does?

Reply:
Hmm, I'm not quite sure I follow - it does basically immediately terminate the writing, since `communicate()` closes stdin, which means that effectively it's sending an EOF to FFmpeg which then terminates writing and finalizes the video file.  So communicate() shouldn't really be waiting for any significant length of time.  In general I would expect close() on a VideoOutput to potentially do some finalization logic that could take a little time, but I can explicitly comment on that or change the method name if you think that's unclear.

Reply:
Got it!